V2 of bulk_upsert and sync

wesleykendall opened the pull request:
@wesokes @jaredlewis @robdmc @thinkt4nk this adds V2 of upsert to manager utils

## Why a second version?

The first version of bulk_upsert has flags for the native upsert and also has multiple parameters for obtaining return values. It is unable to return multi-variable values and categorize them by created/updated though. Along with that, the previous version would not use the native `returning` syntax and it is rather inefficient.

Instead of breaking the interface now, I thought it would be better to migrate things slowly to the newer manager utils interface. 

## What's in the new interface

1. Everything is native except for syncing, which is still done with an additional delete operation. However, using the native return syntax makes syncing much faster, especially if one wants to return values from the sync operation
2. Deadlocks are better prevented for concurrent upserts by sorting the data. Happy to explain offline how this solves deadlock issues
3. Return values are categorized by create / update / delete without doing additional queries. This is in constrast to the previous interface that had an additional restricted option to categorize return values. Sync was not previously capable of this
4. Sync and bulk_upsert have a `returning` argument that defaults to `False`. `True` will return everything, but one can also give a list of fields to `returning` to only return certain fields
5. Previously the interface defaulted to not updating fields unless one specified fields. This previous interface tripped up a lot of people I know that were using it. By default, all applicable fields are updated now unless one specifies the fields to be updated or gives an empty list (e.g. a bulk create that ignores anything that exists). This seems more in line with what one would expect when doing a bulk upsert

## Future plans
Eventually phase out the old `bulk_upsert` and `sync` methods and replace them with the new methods in a manager_utils 2 release

## Testing
I tested this by first duplicating all of the old sync / bulk upsert tests and running them against the new interface. And then I made a couple more tests for returning results

wesokes commented on manager_utils/upsert.py:
```diff
@@ -0,0 +1,224 @@
+    pk_field = model._meta.pk.name
+    if sync:
+        orig_ids = frozenset(queryset.values_list(pk_field, flat=True))
+        deleted = list(orig_ids - frozenset([getattr(r, pk_field) for r in upserted]))
```
any reason to do a frozenset of a list rather than a set comprehension? something like ```{getattr(r, pk_field) for r in upserted}```

also, would it be better to leave the diff as a set rather than converting it to a list?

wesleykendall replied:
@wesokes nope, looks like that was just some old code that was copied and pasted from a long time ago. I updated it to remove the unnecessary list at the end and did a set comprehension

wesokes commented:
pr looks good, just a couple formatting things to stay consistent with the rest of the code:
- for doc strings comments, use a new line after the opening """
- for multiline argument function calls or definitions, if it can't fit on one line, try either
```
1.
def function(
    arg, arg, arg, ...
):
2.
def function(
    arg, arg, arg, ...
    arg, arg, arg, ...
):
3. or one per line, if it helps the readability of a really long argument list
```

wesleykendall commented:
@wesokes addressed comments in https://github.com/ambitioninc/django-manager-utils/pull/102/commits/088e39ce6f5f10276b0bcac652df43b5ae3f70b9

I also built docs locally and verified they rendered correctly. I had to add the napoleon extension to the sphinx config for the google docstrings to render properly. Everything looks good though!
